src/doc/algorithm/shift_image.cpp

```diff
@@ -20,31 +23,111 @@
 namespace doc {
 namespace algorithm {
 
-void shift_image_with_mask(Image* image, const Mask* mask, int dx, int dy)
+void shift_image(Image* image, int dx, int dy, double angle)
 {
-  gfx::Rect bounds = mask->bounds();
-  if (bounds.isEmpty())
-    return;
-
-  // To simplify the algorithm we use a copy of the original image, we
-  // could avoid this copy swapping rows and columns.
-  ImageRef crop(crop_image(image, bounds.x, bounds.y, bounds.w, bounds.h,
-                           image->maskColor()));
-
-  int u = dx;
-  int v = dy;
-  while (u < 0) u += bounds.w;
-  while (v < 0) v += bounds.h;
-
-  for (int y=0; y<bounds.h; ++y) {
-    for (int x=0; x<bounds.w; ++x) {
-      put_pixel(image,
-                bounds.x + ((u + x) % bounds.w),
-                bounds.y + ((v + y) % bounds.h),
-                get_pixel(crop.get(), x, y));
+  if (cos(angle) < -sqrt(2)/2) {
+    dx = -dx;
+    dy = -dy;
+  }
+  else if (sin(angle) >= sqrt(2)/2) {
+    double aux;
+    aux = dx;
+    dx = -dy;
+    dy = aux;
+  }
+  else if (sin(angle) < -sqrt(2)/2) {
+    double aux;
+    aux = dx;
+    dx = dy;
+    dy = -aux;
+  }
+  Image* imageTmp = Image::create(image->pixelFormat(), image->bounds().w, image->bounds().h);
+  imageTmp->copy(image, gfx::Clip(0, 0, 0, 0, image->bounds().w, image->bounds().h));
+
+  for (int y=0; y<image->bounds().h; ++y) {
+    for (int x=0; x<image->bounds().w; ++x) {
+        put_pixel(image,
+                (image->bounds().w + dx + x) % image->bounds().w,
+                (image->bounds().h + dy + y) % image->bounds().h,
+                get_pixel(imageTmp, x, y));
     }
   }
 }
 
+void shift_image_with_mask(Cel* cel, const Mask* mask, int dx, int dy)
+{
+  ASSERT(!cel->bounds().isEmpty());
+  ASSERT(!mask->bounds().isEmpty());
+
+  // Saving initial cel->position() usefull at the end of this function to adjust the cel bounds in shrink process.
+  gfx::Point initialCelPos(cel->position());
+
+  // Making a image which bounds are equal to the UNION of cel->bounds() and mask->bounds()
+  int minX = MIN(cel->bounds().x, mask->bounds().x);
+  int minY = MIN(cel->bounds().y, mask->bounds().y);
+  int maxX = MAX(cel->bounds().x+cel->bounds().w, mask->bounds().x+mask->bounds().w);
+  int maxY = MAX(cel->bounds().y+cel->bounds().h, mask->bounds().y+mask->bounds().h);
+  gfx::Rect compCelBounds = gfx::Rect(minX, minY, maxX - minX, maxY - minY);
```

Comment:
The union of two rectangles can be calculated with the [`gfx::Rect::operator|()`](https://github.com/aseprite/laf/blob/eee4fa27f0f13f7ee03a98283cc41ce1d49c6aa3/gfx/rect.h#L328) (which is a shortcut for [`gfx::Rect::createUnion`](https://github.com/aseprite/laf/blob/eee4fa27f0f13f7ee03a98283cc41ce1d49c6aa3/gfx/rect.h#L247)):
```c++
gfx::Rect compCelBounds = cel->bounds() | mask->bounds();
```
In that way I think you can remove `minX/Y` and `maxX/Y` variables.